### ISSUE TITLE
Google Maps API fails when loaded from file system

### ISSUE THREAD

**Issue by Haraldson** [chentsulin/electron-react-boilerplate#455]
With a packaged production app, the express/hmr server is gone, and everything’s served from the file system. The Google Maps API takes offense, but I imagine other, similar APIs behaves the same way.

Is there a way to get around this without adding a server to the production build? And if not, how would the setup/config look like for the slimmest possible production server?

**Comment by amilajack** [comment]
> Is there a way to get around this without adding a server to the production build? 

Yes. That is a totally viable solution. Just make sure to include express in prod and write a script that will serve the resource/api through http.

**Comment by Haraldson** [comment]
Not sure how to serve/’mask’ only the resources needed, without then actually just serving everything with a server? Isn‘t everything in “file mode” because of `index.html` itself being served from the file system?

Would you guys consider documenting this to some extent? Not Google Maps API specific, but more generally on how to deal with this sort of issue?

**Comment by amilajack** [comment]
Here's a quick example: https://github.com/amilajack/popcorn-time-desktop/blob/09fa99a8f2952a0bdee3b1e881d0f4b4b098fc44/app/api/Subtitle.js#L15-L23.

**Comment by Haraldson** [comment]
Not really able to wrap my head around everything that goes on in that code, and how it will make Google believe that it’s being requested from an http server, and not from the file system?

**Comment by amilajack** [comment]
Those lines ^^ serve static files. Make sure that you change the path that is being served and it should work.

**Comment by amilajack** [comment]
@Haraldson Did everything work out?

**Comment by Haraldson** [comment]
I ended up with a prepackage script that curls the API script URL and saves it to a file on the local file system. It seems like having both `index.html` and the script file being served from the file system does the trick.

**Comment by amilajack** [comment]
🎉

**Comment by Haraldson** [comment]
The hack failed, as Google serves their scripts with some kind of generated token baked in, which times out after some time. Also, serving such an script file through a proxy server such as `express-http-proxy` does nothing, as long as `index.html` itself is served from the file system.

Would very much like to learn how to properly solve loading externally hosted scripts in a packaged app – this can’t be such a special use case..?

**Comment by amilajack** [comment]
Here's how I solved this issue in my project. My app downloads subtitle files that are to be provided to the video element. Video elements don't support the 'file://' protocol. Downloading the subtitle files and moving them to a folder that I served using express turned out to be a pretty good solution for me. I just needed to swap out the `file://Users/...` with something like `http://localhost/subtitleFolderLocation/fileHere`

**Comment by Haraldson** [comment]
I’ve dug a bit deeper now, and it might be some weirdness going on with `webpack.DefinePlugin` making the default Google API key appear in `process.env` – I’ve now tried ‘everything’, but I can’t for the life of me define and overwrite the default, hard-coded key that ships with Electron...

**Comment by amilajack** [comment]
> I can’t for the life of me define and overwrite the default, hard-coded key that ships with Electron

Can you point to some code so I can understand this more? I don't know which environmental variables you're referring to.

**Comment by Haraldson** [comment]
[This is the environment variable I need to set](https://github.com/electron/electron/blob/d7934e7525d2a6344ed16c9506c6a0a52c58a794/docs/api/environment-variables.md#google_api_key), but right now I can’t even seem to pass in the `NODE_ENV` environment variable. It was passing in whatever was in my `~/.bash_profile` for a while, until I deleted that config, as I want it to be different in different environments, naturally.

Basically, this isn’t made available in the app, neither in development or production;
```javascript
new webpack.DefinePlugin({
    'process.env.NODE_ENV': JSON.stringify('production'),
    'process.env.GOOGLE_API_KEY': JSON.stringify('abc123')
})
``
...and I end up with the hard-coded `GOOGLE_API_KEY`.

**Comment by amilajack** [comment]
Here's the `DefinePlugin` example from the docs:

```js
new webpack.DefinePlugin({
    PRODUCTION: JSON.stringify(true),
    VERSION: JSON.stringify("5fa3b9"),
    BROWSER_SUPPORTS_HTML5: true,
    TWO: "1+1"
})
```

A quick way of checking environmental variables is going to the devtools and typing `process.env`.

**Comment by Haraldson** [comment]
[This](https://github.com/chentsulin/electron-react-boilerplate/blob/master/webpack.config.production.js#L51-L53) is how environment variables are set, or are attempted to be set, in this project.

And yes, I’m checking whether the variables are set using the devtools. They’re not being set.

**Comment by amilajack** [comment]
Hmm. Try adding the env variable to the package script:

```js
scripts: {
  package: "cross-env NODE_ENV=production GOOGLE_API_KEY=abc123..."
}
```

**Comment by Haraldson** [comment]
Seeing as I have `NODE_ENV=production` there already, and that’s not available in the running app, I didn’t test doing that before now. No difference, though. Nothing seems to take, somehow.

**Comment by amilajack** [comment]
Do you get something like this:

<img width="452" alt="screen shot 2016-10-24 at 9 42 33 am" src="https://cloud.githubusercontent.com/assets/6374832/19654683/3ba0e836-99ce-11e6-92fe-43a11d7ba949.png">

**Comment by Haraldson** [comment]
Yup.

**Comment by amilajack** [comment]
How about something like this:

```js
new webpack.DefinePlugin({
    'process.env.NODE_ENV': JSON.stringify('production'),
    'process.env.GOOGLE_API_KEY': process.env.CUSTOM_GOOGLE_API_KEY
})
```

```
scripts: {
  package: "cross-env NODE_ENV=production CUSTOM_GOOGLE_API_KEY=abc123..."
}
```

**Comment by Haraldson** [comment]
Nope. The custom key isn’t available in `process.env` in the running app, whereas `GOOGLE_API_KEY ` remains the hardcoded one that ships with every Electron instance.

**Comment by amilajack** [comment]
So if I understand you correctly, `process.env.CUSTOM_GOOGLE_API_KEY` or any other environmental variable cannot be added to electron?

**Comment by Haraldson** [comment]
That seems to be the case... :-S

**Comment by amilajack** [comment]
But this does work?:

```js
new webpack.DefinePlugin({
    'process.env.NODE_ENV': JSON.stringify('production'),
    'process.env.GOOGLE_API_KEY': JSON.stringify('abc123')
})
```

**Comment by Haraldson** [comment]
Neither of these two ways of defining environment variables work... This is so frustrating.

**Comment by Haraldson** [comment]
I’m able to get the correct environment variables if I do

`cross-env NODE_ENV=production GOOGLE_API_KEY=abc123 open release/my.app`

But why aren’t they persisted somewhere?

**Comment by amilajack** [comment]
> This is so frustrating.

Yes, it is. I've had a bunch of issues with electron and its internals myself.

I'm not sure if this helps but here's how I handle environment configuration:
[Use EnvironmentPlugin to inject vars](https://github.com/amilajack/popcorn-time-desktop/blob/09fa99a8f2952a0bdee3b1e881d0f4b4b098fc44/webpack.config.renderer.js#L12)
[Inject env variables in `.env` in development](https://github.com/amilajack/popcorn-time-desktop/blob/09fa99a8f2952a0bdee3b1e881d0f4b4b098fc44/main.development.js#L24)

**Comment by Haraldson** [comment]
Tried pretty much copying your approach now, after having tried just using EnvironmentPlugin in combination with DefinePlugin earlier. No difference.

```javascript
// Get all the possible flags
import fs from 'fs'
import dotenv from 'dotenv'

const environmentVariables = dotenv.parse(new Buffer(fs.readFileSync('.env.development', { encoding: 'utf8' })))
const flags = Object.keys(environmentVariables)

...

plugins: [
    new webpack.EnvironmentPlugin(flags),
    new webpack.DefinePlugin({ 'process.env': JSON.stringify({ ...process.env, ...environmentVariables }) })
]
```

If I dump the result of `{ ...process.env, ...environmentVariables }`, all the variables are there, so they disappear for some reason, at some point. In development mode, `NODE_ENV` is correctly set to `'development'`, but the default `GOOGLE_API_KEY` is still the one I get there, too.

Any ideas on what could cause the variables to disappear along the way?

**Comment by Haraldson** [comment]
OK, so my cumbersome solution is this postpackage script:

```javascript
'use strict'

import fs from 'fs'
import plist from 'plist'
import dotenv from 'dotenv'

export default (options) => {
    if(options.platform !== 'darwin')
        return

    // Get environment variables
    const env = dotenv.parse(new Buffer(fs.readFileSync('.env.production', { encoding: 'utf8' })))

    // Generate path to plist
    const { dir, out, name, platform, arch } = options
    const plistPath = `${dir}${out}/${name}-${platform}-${arch}/${name}.app/Contents/Info.plist`

    // Parse plist to plain
    let plistParsed = plist.parse(fs.readFileSync(plistPath, 'utf8'))

    // Add environment variables
    plistParsed.LSEnvironment = env

    // Write everything back to plist
    fs.writeFileSync(plistPath, plist.build(plistParsed), 'utf8')
}
```

**Comment by amilajack** [comment]
So everything works now?

**Comment by Haraldson** [comment]
This makes everything work, however it does feel like a big hack though.

**Comment by amilajack** [comment]
Yea it does :( sorry I couldnt be of much help with this issue.
